Add high-performance timeseries anomaly detection and imputation module

ezwelty opened the pull request:
The `pudl.analysis.demand_anomalies` module contains a `Series` class representing a generic multivariate data series. It has three main components:

- A fast implementation of the anomaly screening methods developed by Tyler Ruggles for EIA demand data (https://github.com/truggles/EIA_Cleaned_Hourly_Electricity_Demand_Code).
- Xinyu Chen's fast and accurate imputation algorithms LATC-TNN and LATC-Tubal (https://github.com/xinychen/tensor-learning/tree/master/mats) which appear to even outperform the MICE method used by Tyler Ruggles. See https://github.com/catalyst-cooperative/pudl/issues/598#issuecomment-752790831.
- A very flexible method for inserting null values for validating imputation results.

Here is an example using Tyler Ruggles' EIA data published at https://zenodo.org/record/3690240.

```python
import glob
import os

import matplotlib.pyplot as plt
import pandas as pd
from pudl.analysis.timeseries_cleaning import Timeseries

# --- Read demand data --- #
paths = glob.glob('data/release_2019_Oct/balancing_authorities/*.csv')
xs = []
for path in paths:
    ba = os.path.splitext(os.path.basename(path))[0]
    df = pd.read_csv(
        path,
        na_values=["MISSING", "EMPTY"],
        usecols=["date_time", "raw demand (MW)"],
        parse_dates=["date_time"]
    ).set_index('date_time')
    xs.append(df[['raw demand (MW)']].rename(columns={'raw demand (MW)': ba}))
x = pd.concat(xs, axis=1)

# --- Flag anomalous values --- #
s = Timeseries(x)
s.flag_ruggles()
# Which is equivalent to:
# s.flag_negative_or_zero()
# s.flag_identical_run()
# s.flag_global_outlier()
# s.flag_global_outlier_neighbor()
# s.flag_local_outlier()
# s.flag_double_delta()
# s.flag_single_delta()
# s.flag_anomalous_region(window=49)

# --- Review and plot flagged values --- #
s.summarize_flags()
s.plot_flags(0)
plt.show()

# --- Validate imputation results --- #
# Null additional values following Tyler Ruggles' method
mask = s.simulate_nulls(overlap=True, intersect=True)
imputed = s.impute(mask=mask, method='tnn', blocks=4)
# Review deviations between original and imputed values
s.summarize_imputed(imputed, mask)
```

zaneselvans reviewed (approved):
Wow, this is a lot. I definitely don't understand what all is going on in tensorland, but from the outside interface, it looks very usable.  Given that the anomaly flagging and imputation functionality are all in a single module, maybe we should pick a more informative name than `demand_anomalies`.

Should we be worried about the `Series` class colliding with `pandas.Series`?

It looks like the CI failure on the `demand-anomalies` branch is just because it doesn't have the (many) changes that have accumulated in `sprint29` over the last few months.

zaneselvans commented:
@ezwelty How hard would it be for you to add some unit tests for this module? We've committed to moving in that direction, and having some examples to learn from would be great -- in addition to ensuring that the module's functionality is tested.

zaneselvans commented:
This will also go a long way toward closing #661. Very curious to see what this all looks like on the FERC 714 data.

ezwelty commented on src/pudl/analysis/demand_anomalies.py:
```diff
@@ -0,0 +1,1316 @@
+# ---- Anomaly detection ---- #
+
+
+class Series:
```
Coincidence. `MultivariateTimeseries` is a mouthful. How about `Timeseries`? I chose `Series` to be more generic, but its uses are likely to be limited to timeseries cleaning.

ezwelty commented on src/pudl/analysis/demand_anomalies.py:
```diff
@@ -0,0 +1,1316 @@
+            mask[shift:][outliers[:-shift]] = True
+        self.flag(mask, "GLOBAL_OUTLIER_NEIGHBOR")
+
+    @functools.lru_cache(maxsize=2)
```
Docs: https://docs.python.org/3/library/functools.html#functools.lru_cache

The "metric" methods (e.g. `rolling_median`, etc) are used by Tyler Ruggles' more complex anomaly detection routines. In some cases, a `flag_*` method needs access to both a metric A and another metric B which depends on metric A. To make this more transparent, I broke out each metric into its own documented method. However, to avoid recomputing the same metric from the same `flag_*` method, I use a cache on the reused metric method, which is then cleared upon the completion of the `flag_*` method.

ezwelty commented:
> Given that the anomaly flagging and imputation functionality are all in a single module, maybe we should pick a more informative name than demand_anomalies.

`timeseries_cleaning` ?

> Should we be worried about the Series class colliding with pandas.Series?

`MultivariateTimeseries` is a mouthful. How about `Timeseries`? I chose `Series` to be more generic, but its uses are likely to be limited specifically to *time*series cleaning. Open to other suggestions.

> How hard would it be for you to add some unit tests for this module? We've committed to moving in that direction, and having some examples to learn from would be great -- in addition to ensuring that the module's functionality is tested.

I could probably write some fast validation tests using something like perturbed sine waves.
Some functions already have doctests in their docstrings. Should I continue in that same vein, and/or add a test module?
If the doctests remain,`pytest` calls should include `--doctest-modules src/pudl` going forward (I also already wrote doctests for my harvest and spatial allocation modules). Doctests could also be moved into the API documentation, but that requires passing on `automodule` and using `autofunction` and `autoclass` instead (e.g. https://github.com/python-attrs/attrs/blob/master/docs/api.rst).

zaneselvans commented:
* Some fast validation tests using sinewaves sound great to me.
* My gut feeling is that we probably want to stick to test modules for the moment, since we haven't been doing doctests anywhere else, and have really struggled to even get folks to write docstrings which have valid formatting as it is.
* I think we *do* want to move toward using static type checking, and right now that's not integrated into any of the tests / linting / CI. Would it be difficult for you to get that nominally setup?

ezwelty commented:
> Some fast validation tests using sinewaves sound great to me.

Added a test module. Very basic, but at least ensures that everything runs and produces reasonable results.

I abstained from attempting unit tests for each flag because the ones for the more complex flags would take some time to design, and as you said elsewhere, the ultimate goal is a more universal anomaly detection which will do away with these individual checks.

> My gut feeling is that we probably want to stick to test modules for the moment.

Sure, but why not include those that already exist and add them to the coverage report?

> I think we do want to move toward using static type checking, and right now that's not integrated into any of the tests / linting / CI. Would it be difficult for you to get that nominally setup?

Setting up the type checking isn't the hard part. It is dealing with the flurry of errors that inevitably arise – some of which inevitably are spurious and can be a pain to deal with.

I've added a basic mypy configuration file and fixed type checking errors in the new modules introduced by this pull request. A humble beginning.

Between type checking and the doctests, maybe we can have a chat at some point about the current hooks/tests/coverage design. Although I already have a lot of things started that I should probably finish first...

> Timeseries and timeseries_cleaning sound good to me.

Ok, done.

zaneselvans commented:
@rousik has a preference for putting the unit test modules next to the modules that they apply to, and he developed the Excel extractor. We were also thinking about putting the integration tests under `test` and the unit tests with the modules, at his suggestion. In the `prefect` branch he's integrated some additional unit test modules into the CI as well.  We do need to have some kind of testing design discussion. There's a lot going on, so yeah probably better to just create an issue or three and finish up the things that are already in progress before starting new threads.

Admittedly, I set up the test coverage (and all of the testing stuff) quite a while ago, and I'm sure that it is not... entirely optimal. Or even necessarily correct. There were some modules that got excluded because they're kind of leftover junk drawers (`analysis.py` :eyes:). We definitely want to encourage > 0% coverage! If you see obvious improvements to make please highlight them! Or make a PR.
